### DISCUSSION TITLE
Add option in settings to display sub-pages in side panel in reverse date order

### DISCUSSION THREAD

**Discussion by enzedonline** [wagtail/wagtail#8586] [Ideas]
Would be good to display sub-pages in the side panel in reverse edited date order to match the listing page. Currently they're listed in order of creation date (oldest at the top). Not so useful when you have something like a blog or news site etc:

![image](https://user-images.githubusercontent.com/72310199/170030530-576085b8-8c70-47bf-b739-e646736632db.png)

![image](https://user-images.githubusercontent.com/72310199/170030624-0f1e1e39-e3c9-45e5-8bf6-7b2b17c40332.png)

Either as default, or as a configurable option in settings would be great.

**Comment by th3hamm0r** [comment]
+1
Or maybe add/improve hooks for that, so you can modify that behavior based on the parent page for example.
_Fyi: Actually the items are sorted by their default ordering (`path`, also used for menus), new pages always get appended to the end. So you can of course change this manually, but in some cases that's not an option (e.g. if a news-index has thousands of child pages)._

For now, if we have to tweak this ordering for a custom `FolderPage` type, we add the following to `wagtail_hooks.py`, but its really hacky and could break easily with new wagtail updates:
```python
@hooks.register('construct_explorer_page_queryset')
def menu_explorer_chronological_folder_page_ordering(parent_page, pages, request):
    # HACK: The page explorer in the content already uses the correct ordering (last edited), but the explorer
    # in the menu uses the menu-ordering. Debugging has shown that the menu-explorer uses the param
    # "for_explorer=1", so we check that too to avoid overriding the content explorer's query.
    # See wagtail.admin.api.filters.ForExplorerFilter.filter_queryset()
    if request.GET.get('for_explorer'):
        if parent_page.content_type_id == ContentType.objects.get_for_model(FolderPage).id:
            # This is also the default ordering in the content's explorer.
            pages = pages.order_by("-latest_revision_created_at")

    return pages


@hooks.register('construct_page_chooser_queryset')
def page_chooser_chronological_folder_page_ordering(pages, request):
    # HACK: Unfortunately this hook does not get the parent page, if it is browsed. Debugging has shown, that
    # there are two views, which call this hook: search and browse. The search view searches throw all pages,
    # so we can only change the browse query, where the parent_page_id-kwarg is set. With this id we can check
    # if this is a folder page and therefore apply the chronological ordering.
    # See wagtail.admin.views.chooser.browse()
    if request.resolver_match and request.resolver_match.url_name == "wagtailadmin_choose_page_child":
        parent_page_id = request.resolver_match.kwargs.get("parent_page_id")
        if parent_page_id and Page.objects.type(FolderPage).filter(id=parent_page_id).exists():
            # This is also the default ordering in the content's explorer.
            pages = pages.order_by("-latest_revision_created_at")

    return pages
```

To avoid those hacks, it would be helpful if:
* `construct_explorer_page_queryset` would get the context it is used in (menu explorer vs. index-view)
* `construct_page_chooser_queryset` also gets its context or is split into more specific hooks

**Reply by enzedonline** [reply]
Would it be simpler to just have a Page attribute that instructs Wagtail how to sort the menu items for that class? Default is as is but can be overridden.

Something along the lines of

```
class NewsItem(Page):
    ....
    menu_sort_order = 'desc_edited'
    ....
```

**Reply by gasman** [reply]
I think this might be a duplicate of #882.

**Comment by phildexter** [comment]
Agree that as it is right now isn't great. 

Interestingly when we've thought about this before we've gone to order first by whether a page has children, and then alphabetically. This tests well with users - the understand the ordering at least.

Though I can see ordering by date being useful for some sites, it might confuse users as we don't currently show the date in that menu - so we would probably want to add that in to give the user a hint of what order they are in, and where in the menu the page they are looking for might be.

It feels right to me that this be something customisable per site, but we'd need to choose a default. I think as a default ordering by  whether a page has children, and then alphabetical, might make the most sense?

**Reply by enzedonline** [reply]
@phildexter yeah, agree on the pages with children coming first, I generally manually order smaller groups of pages this way.

Maybe what's needed is a union of 2 queries, something like:
  pages with children, alphabetically (?)
  pages without children, customizable

After a little more thought, it makes more sense to me to make the custom sort order at the parent page class with something like child_page_menu_order as an attribute of Page :

```
class NewsListPage(Page)
  ...
  child_page_menu_order = '-creation_date'
  ...
```

Say there's a magazine list child page and a editorial-list child page, these would appear at the top ignoring the custom ordering, then the news articles (which don't have child pages) would follow in reverse creation order.

I guess another option is to set the menu order from a session cookie, whichever way I've chosen to order the pages in the main list for that parent page is reflected in the menu (if that makes sense).

**Reply by phildexter** [reply]
Yeah I really like the idea that it could be set at a page type level - probably with the default being alphabetical, but easily switchable to creation date or last edited date.

I think this would / should also be carried through to the page listing view as well, so the menu and listing pages are always in the same order (as per your initial comment!)
